contracts/cw20-staking/src/error.rs

```diff
@@ -0,0 +1,60 @@
+use cosmwasm_std::{StdError, Uint128};
+use thiserror::Error;
+
+#[derive(Error, Debug)]
+pub enum ContractError {
+    #[error("{0}")]
+    Std(#[from] StdError),
+
+    #[error("Unauthorized")]
+    Unauthorized {},
+
+    #[error("Validator '{0}' not in current validator set")]
+    NotInValidatorSet(String),
+
+    #[error("Different denominations in bonds: '{0}' vs. '{1}'")]
+    DifferentBondDenom(String, String),
+
+    #[error("Stored bonded {0}, but query bonded {1}")]
+    BondedMismatch(Uint128, Uint128),
+
+    #[error("No {0} tokens sent")]
+    EmptyBalance(String),
+
+    #[error("Must unbond at least {0} {1}")]
+    UnbondTooSmall(Uint128, String),
+
+    #[error("Insufficient balance in contract to process claim")]
+    BalanceTooSmall {},
+
+    #[error("Cannot set to own account")]
+    CannotSetOwnAccount {},
+
+    #[error("Invalid zero amount")]
+    InvalidZeroAmount {},
+
+    #[error("Allowance is expired")]
+    Expired {},
+
+    #[error("No allowance for this account")]
+    NoAllowance {},
+
+    #[error("Minting cannot exceed the cap")]
+    CannotExceedCap {},
+}
+
+impl From<cw20_base::ContractError> for ContractError {
+    fn from(err: cw20_base::ContractError) -> Self {
```

Comment:
Nice extension.

I did try to figure out how to "extend" enums in rust and it is not possible.
I wonder if we could simplify this one day with codegen? But I guess it is only a few minutes to write such code now.

Reply:
Yeah, these cases made me search for "rust inheritance" and the like... traits are fine and dandy, but for abstraction there's nothing better than subtypes (interface inheritance) IMO.

Reply:
Come to think of it, this is probably more a case of requiring "plain old" (implementation) inheritance, rather than interface inheritance.